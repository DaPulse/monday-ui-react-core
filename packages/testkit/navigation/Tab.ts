import { Page, Locator, test } from "@playwright/test";
import { BaseElement } from "../BaseElement";

/**
 * Class representing a Tab element.
 * Extends the BaseElement class.
 */
export class Tab extends BaseElement {
  /**
   * Create a Tab.
   * @param {Page} page - The Playwright page object.
   * @param {Locator} locator - The locator for the Tab element.
   * @param {string} elementReportName - The name for reporting purposes.
   */
  constructor(page: Page, locator: Locator, elementReportName: string) {
    super(page, locator, elementReportName);
  }

  /**
   * Click the Tab element.
   * @returns {Promise<void>}
   */
  async click(): Promise<void> {
    await test.step(`Click on ${this.elementReportName} tab`, async () => {
      await this.locator.click();
    });
  }
}
