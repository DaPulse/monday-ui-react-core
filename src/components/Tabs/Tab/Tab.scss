@import "../../../styles/themes";
@import "../../../styles/typography";

.tab--wrapper {
  display: inline-block;
  border-bottom: 2px solid;
  @include theme-prop(border-color, primary-background-hover-color);
  text-align: center;
  height: 100%;

  .tab-inner {
    $margin-bottom: 1px;
    height: calc(100% - #{$margin-bottom});
    display: flex;
    align-items: center;
    justify-content: center;
    text-decoration: none;
    padding: 4px 16px;
    @include theme-prop(color, primary-text-color);
    font-weight: 400;
    font-size: 16px;
    user-select: none;
    margin-bottom: $margin-bottom;
    cursor: pointer;

    &:focus {
      outline: none;
    }
  }

  &.active {
    @include theme-prop(border-bottom-color, primary-color);
  }

  &.disabled {
    @include theme-prop(color, disabled-text-color);

    .tab-inner {
      cursor: not-allowed;
    }
  }

  &:not(.disabled) {
    .tab-inner {
      &:hover {
        border-radius: 4px;
        @include theme-prop(background-color, primary-background-hover-color);
      }
    }
  }
}
