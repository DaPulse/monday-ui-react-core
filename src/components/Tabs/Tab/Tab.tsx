/* eslint-disable jsx-a11y/anchor-is-valid,jsx-a11y/click-events-have-key-events */
import cx from "classnames";
import React, { FC, forwardRef, ReactElement, useRef } from "react";
import { noop as NOOP } from "lodash-es";
import useMergeRefs from "../../../hooks/useMergeRefs";
import Icon, { IconSubComponentProps } from "../../Icon/Icon";
import VibeComponentProps from "../../../types/VibeComponentProps";
import { IconType } from "../../Icon/IconConstants";
import { ComponentDefaultTestId, getTestId } from "../../../tests/test-ids-utils";
import styles from "./Tab.module.scss";

export interface TabProps extends VibeComponentProps {
  /**
   * Class name for tab link-name
   */
  tabInnerClassName?: string;
  /**
   * Tab index
   */
  value?: number;
  disabled?: boolean;
  active?: boolean;
  focus?: boolean;
  icon?: string | React.FunctionComponent<IconSubComponentProps> | null;
  iconType?: IconType;
  iconSide?: string;
  onClick?: (value: number) => void;
  /**
   * Tab link-name
   */
  children?: string | ReactElement[];
}

const Tab: FC<TabProps> = forwardRef(
  (
    {
      className,
      tabInnerClassName,
      id,
      value = 0,
      disabled = false,
      active = false,
      focus = false,
      onClick = NOOP,
      icon,
      iconType,
      iconSide = "left",
      children,
      "data-testid": dataTestId
    },
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });

    function renderIconAndChildren() {
      if (!icon) return children;

      const iconElement = (
        <Icon
          clickable={false}
          ariaHidden={true}
          iconType={iconType}
          icon={icon}
          // css-modules-migration: iconSide - used only for overrides
          className={cx(styles.tabIcon, "tab-icon", iconSide)}
          iconSize={18}
          ignoreFocusStyle
        />
      );

      if (iconSide === "left") {
        return [iconElement, ...children];
      }

      return [...children, iconElement];
    }
    return (
      <li
        ref={mergedRef}
        key={id}
        className={cx(styles.tabWrapper, "tab--wrapper", className, {
          [styles.active]: active,
          ["active"]: active,
          [styles.disabled]: disabled,
          ["disabled"]: disabled,
          [styles.tabFocusVisibleInset]: focus,
          ["tab-focus-visible-inset"]: focus
        })}
        id={id}
        role="tab"
        aria-selected={active}
        aria-disabled={disabled}
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.TAB, id)}
      >
        <a className={cx(styles.tabInner, "tab-inner", tabInnerClassName)} onClick={() => !disabled && onClick(value)}>
          {renderIconAndChildren()}
        </a>
      </li>
    );
  }
);

export default Tab;
