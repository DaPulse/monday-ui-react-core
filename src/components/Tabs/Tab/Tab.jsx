import React, { useRef, forwardRef } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import useMergeRefs from "../../../hooks/useMergeRefs";
import "./Tab.scss";

const Tab = forwardRef(({ className, id, value, disabled, active, onClick, children }, ref) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });

    return (
      <li ref={mergedRef}
          key={id}
          className={cx("tab--wrapper", className, { active, disabled })}
          id={id}
          aria-selected={active}
          aria-disabled={disabled}
      >
          <a role="tab"
             className="tab-inner"
             onClick={() => !disabled && onClick(value)}
          >
              {children}
          </a>
      </li>
  );
});

Tab.propTypes = {
    className: PropTypes.string,
    id: PropTypes.string,
    value: PropTypes.number,
    disabled: PropTypes.bool,
    active: PropTypes.bool,
    onClick: PropTypes.func
};
Tab.defaultProps = {
    className: "",
    id: "",
    value: 0,
    disabled: false,
    active: false
};

export default Tab;
