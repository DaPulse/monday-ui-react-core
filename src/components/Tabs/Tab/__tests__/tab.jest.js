import React from "react";
import renderer from "react-test-renderer";
import Tab from "../Tab";

it("renders correctly with empty props", () => {
  const tree = renderer.create(<Tab />).toJSON();
  expect(tree).toMatchSnapshot();
});
