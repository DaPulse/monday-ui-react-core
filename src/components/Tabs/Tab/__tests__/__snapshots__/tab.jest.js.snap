// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with empty props 1`] = `
<li
  aria-disabled={false}
  aria-selected={false}
  className="tab--wrapper"
  id=""
  role="tab"
>
  <a
    className="tab-inner"
    onClick={[Function]}
  />
</li>
`;
